SignatureHash(): Document buggy but consensus-critical legacy behavior, aid in preventing a similar fate for BCH signature hashing

awemany opened the pull request:
As discussed previously, I was unhappy about the implicit failure of the SignatureHash(..) function and returning a simple value of 1 without any further, local handling of that value.

This change makes this failure handling more explicit and adds a notice to the docs of SignatureHash(..) to be careful when using it, in addition to hiding the implementing subroutines that were exposed before.

As per my limited understanding, ECDSA signing assumes in its security model that a cryptographic hash of the message is signed, always. The current implementation allows to call SignatureHash(..) with invalid parameters, resulting in a return value 1 *and does further signing and processing of this resulting error code*.

This needs careful review as this touches some key areas.

As far as I understand, returning false in TransactionSignatureCreator::CreateSig(..) is the right response and indicates failure to the higher layers on top of that method, but I might be wrong here and again like someone else to check.

@gandrewstone : I prepared the PR now to keep track of this but I can of course rebase later on top of your changes. When this is ready to be merged at some point in the future, feel free to squash it all together, of course.

awemany committed:
Check SignatureHash(..) return value in unit tests and benchmarks

This renames the 'one' uint256(1) value to SIGNATURE_HASH_ERROR and exports
it from interpreter.h. It further adds checks for this value where it should
not happen to the unit tests and the benchmark.

gandrewstone commented:
makes sense to me.  There is a 1 in 2^256 chance that data will actually hash to uint256(1).  I think that this is acceptable, esp since the only effect will be an inability to spend that exact transaction.  Any slight change to the tx will make it spendable.

awemany commented:
@gandrewstone: Alright, thanks for the feedback. Another option would be to use an even more standard way to signal error conditions with something like an additional `bool* perror` flag or so. What do you think?
I did like i did because I wanted to do minimum amount of changes, but it seems like that would be even cleaner.

awemany commented:
Taking this out of WIP. A couple further notices: This odd behavior of the SignatureHash() function is actually (due to a further missing check in CheckSig) a known cause of consensus-critical bugs in Bitcoin Legacy!

So the commits have correspondingly been changed to remove uint256(1)-is-error-signalling from the BCH signature function as well as properly documenting this broken behavior.

sickpig commented on src/script/interpreter.cpp:
```diff
@@ -1526,18 +1526,85 @@ class CTransactionSignatureSerializer
         //  nIn out of range
-        return one;
+
+        // Important notice:
```
Use uppercase chars?

awemany committed:
cashlib: Use generic SignatureHash()

Use the generic SignatureHash() function instead of the
SignatureHashBitcoinCash() implementation detail.

Also enforce / check for the SIGHASH_FORKID bit being set using a
a DbgAssert(..).

awemany committed:
Move all SignatureHash..() to sigcommon.cpp

And also unexport the specialized functions from interpreter.h, to only
present the unified SignatureHash()-interface.

awemany committed:
Add notices about returning uint256(1) to SignatureHashLegacy(..)

This adds a couple warnings to the SignatureHashLegacy() function
that returning uint256(1) one to signal errors isn't the best thing
to do and should not repeated in any future hash function implementations.

awemany committed:
SignatureHashBitcoinCash(): Never call without SIGHASH_FORKID

SignatureHashBitcoinCash() was never called without the fork id flag
set (as it was only called from the SignatureHash() function).

Using a special value of uint256(1) to signal an error is breaking assumptions
about ECDSA signing and this mistake should not be repeated for the Bitcoin Cash
signature function.

This removes the return-1 code-path from SignatureHashBitcoinCash
as well as adding a warning that this function is not to be used ever without
SIGHASH_FORKID set.

awemany committed:
CheckSig: Remove dependency on SignatureHashBitcoinCash()

Calling SignatureHashBitcoinCash() is implicit if the SIGHASH_FORKID
flag is set. This removes the explicit dependency on
SignatureHashBitcoinCash() and replaces it with a generic SignatureHash()
call.

Also removes the SignatureHashBitcoinCash() export.

sickpig commented on src/script/sigcommon.cpp:
```diff
@@ -53,68 +53,233 @@ uint256 GetOutputsHash(const CTransaction &txTo)
-        uint256 hashSequence;
-        uint256 hashOutputs;
+        //  nIn out of range
+        // Important notice:
```
Saw you moved the code under `sigcommon.cpp`, still I'd put this comment line in upper case.

awemany replied:
Alright, fixed.

gandrewstone commented on src/cashlib/cashlib.cpp:
```diff
@@ -154,7 +157,7 @@ extern "C" int SignTx(unsigned char *txData,
     CKey key = LoadKey(keyData);
 
     size_t nHashedOut = 0;
-    uint256 sighash = SignatureHashBitcoinCash(priorScript, tx, inputIdx, nHashType, inputAmount, &nHashedOut);
```
why not use SignatureHashBitcoinCash here?

awemany replied:
Well, my general guideline was to always try to reduce the surface area and thus coupling between the different parts of `bitcoind`. What would be the reason to be more specific here?

gandrewstone replied:
this change increases the surface area.  SignatureHash calls either ...Legacy or ...BitcoinCash depending on flags.  I wanted to restrict cashlib to BitcoinCash signature hash only to have one less thing that can go wrong for the wallet dev (accidentally Legacy signing in the BitcoinCash network)...
